test/cufinufft3d3_test.cu

```diff
@@ -0,0 +1,228 @@
+	checkCudaErrors(cudaMemcpy(d_u,u,N*sizeof(FLT),cudaMemcpyHostToDevice));
+	checkCudaErrors(cudaMemcpy(d_c,c,M*sizeof(CPX),
+		cudaMemcpyHostToDevice));
+
+	cudaEvent_t start, stop;
+	float milliseconds = 0;
+	float totaltime = 0;
+	cudaEventCreate(&start);
+	cudaEventCreate(&stop);
+
+    // warm up CUFFT (is slow, takes around 0.2 sec... )
+	cudaEventRecord(start);
+	{
+		int nf1=1;
+		cufftHandle fftplan;
+		cufftPlan1d(&fftplan,nf1,CUFFT_TYPE,1);
+	}
+	cudaEventRecord(stop);
+	cudaEventSynchronize(stop);
+	cudaEventElapsedTime(&milliseconds, start, stop);
+	printf("[time  ] dummy warmup call to CUFFT\t %.3g s\n", milliseconds/1000);
+
+        // now to the test...
+	CUFINUFFT_PLAN dplan;
+	int dim = 3;
+	int type = 3;
+
+	// Here we setup our own opts, for gpu_method.
+	cufinufft_opts opts;
+	ier=CUFINUFFT_DEFAULT_OPTS(type, dim, &opts);
+	if(ier!=0){
+	  printf("err %d: CUFINUFFT_DEFAULT_OPTS\n", ier);
+	  return ier;
+	}
+	opts.gpu_method=method;
+
+	int ntransf = 1;
+	int maxbatchsize = 1;
+
+	cudaEventRecord(start);
+	{
+		PROFILE_CUDA_GROUP("cufinufft3d_plan",2);
+		ier=CUFINUFFT_MAKEPLAN(type, dim, NULL, iflag, ntransf, tol,
+				       maxbatchsize, &dplan, &opts);
+		if (ier!=0){
+			printf("err: cufinufft_makeplan\n");
+		}
+	}
+	cudaEventRecord(stop);
+	cudaEventSynchronize(stop);
+	cudaEventElapsedTime(&milliseconds, start, stop);
+	totaltime += milliseconds;
+	printf("[time  ] cufinufft plan:\t\t %.3g s\n", milliseconds/1000);
+
+	cudaEventRecord(start);
+	{
+		PROFILE_CUDA_GROUP("cufinufft_setpts",3);
+		ier=CUFINUFFT_SETPTS(M, d_x, d_y, d_z, N, d_s, d_t, d_u, dplan);
+		if (ier!=0){
+		  printf("err: cufinufft_setpts\n");
+		  return ier;
+		}
+	}
+	cudaEventRecord(stop);
+	cudaEventSynchronize(stop);
+	cudaEventElapsedTime(&milliseconds, start, stop);
+	totaltime += milliseconds;
+	printf("[time  ] cufinufft setNUpts:\t\t %.3g s\n", milliseconds/1000);
+
+	cudaEventRecord(start);
+	{
+		PROFILE_CUDA_GROUP("cufinufft_execute",4);
+		ier=CUFINUFFT_EXECUTE(d_c, d_fk, dplan);
+		if (ier!=0){
+		  printf("err: cufinufft_execute\n");
+		  return ier;
+		}
+	}
+	cudaEventRecord(stop);
+	cudaEventSynchronize(stop);
+	cudaEventElapsedTime(&milliseconds, start, stop);
+	totaltime += milliseconds;
+	float exec_ms =	milliseconds;
+	printf("[time  ] cufinufft exec:\t\t %.3g s\n", milliseconds/1000);
+
+	cudaEventRecord(start);
+	{
+		PROFILE_CUDA_GROUP("cufinufft3d_destroy",5);
+		ier=CUFINUFFT_DESTROY(dplan);
+		if (ier!=0){
+		  printf("err: cufinufft_destroy\n");
+		  return ier;
+		}
+	}
+	cudaEventRecord(stop);
+	cudaEventSynchronize(stop);
+	cudaEventElapsedTime(&milliseconds, start, stop);
+	totaltime += milliseconds;
+	printf("[time  ] cufinufft destroy:\t\t %.3g s\n", milliseconds/1000);
+
+	checkCudaErrors(cudaMemcpy(fk,d_fk,N*sizeof(CUCPX),cudaMemcpyDeviceToHost));
+
+	printf("[Method %d] %ld U pts to %d NU pts in %.3g s:\t%.3g NU pts/s\n",
+			opts.gpu_method, N ,M,totaltime/1000,M/totaltime*1000);
```

Comment:
For timings in type 3 in CPU code we report (M+N)/totaltime   not M/totaltime.     Ie use the *total* N+M   instead of M each time it appears here. It would be good to make this match the CPU code otherwise we're off by factors of two when M=N, etc...  Thanks, Alex

Reply:
Makes sense. I've changed it to `M+N`.